internal/graph/check_test.go

```diff
@@ -581,3 +578,175 @@ condition condition1(param1: string) {
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cyclicErrorHandler, falseHandler)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if_a handler_errors_with_cycle_but_other_handler_is_truthy,_return_allowed:true_with_a_nil_error", func(t *testing.T) {
+		cyclicErrorHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrCycleDetected
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cyclicErrorHandler, trueHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+
+	t.Run("should_aggregate_DatastoreQueryCount_of_non-error handlers", func(t *testing.T) {
+		trueHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			time.Sleep(5 * time.Millisecond) // forces `trueHandler` to be resolved after `falseHandler`
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(1),
+				},
+			}, nil
+		}
+
+		falseHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(5),
+				},
+			}, nil
+		}
+
+		errorHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(9999999),
+				},
+			}, ErrResolutionDepthExceeded
+		}
+
+		resp, err := union(ctx, concurrencyLimit, falseHandler, trueHandler, errorHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+		require.Equal(t, uint32(5+1), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should_aggregate_DatastoreQueryCount_of_all_falsy_handlers", func(t *testing.T) {
+		handler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(3),
+				},
+			}, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, handler, handler, handler) // three handlers
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+		require.Equal(t, uint32(3*3), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should_return_allowed:true_if_truthy_handler_evaluated_before_handler_cancels_via_context", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+		t.Cleanup(cancel)
+
+		resp, err := union(ctx, concurrencyLimit, trueHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+	t.Run("should_handle_cancellations_through_context", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+		t.Cleanup(cancel)
+
+		slowHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			time.Sleep(50 * time.Millisecond)
+			return nil, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, slowHandler, falseHandler)
```

Comment:
do you need to pass the falseHandler? i expect that because of the deadline exceeded we are already setting false

Reply:
Not necessarily but just serves to demonstrate that you could pass another deterministic handler and it wouldn't affect the outcome; the union waits for the `slowHandler`.